docs/frontend-system/utility-apis/05-migrating.md

```diff
@@ -0,0 +1,138 @@
+
+## React package interface and ref changes
+
+Let's begin with [your `-react` package](../../architecture-decisions/adr011-plugin-package-structure.md). The act of exporting TypeScript interfaces and API refs have not changed from the old system. You can typically keep those as-is. For illustrative purposes, this is an example of an interface and its API ref:
+
+```tsx title="in @internal/plugin-example-react"
+import { createApiRef } from '@backstage/frontend-plugin-api';
+
+/**
+ * Performs some work.
+ * @oublic
+ */
+export interface WorkApi {
+  doWork(): Promise<void>;
+}
+
+/**
+ * The work interface for the Example plugin.
+ * @public
+ */
+export const workApiRef = createApiRef<WorkApi>({
+  id: 'plugin.example.work',
+});
+```
+
+In this example, the plugin ID already follows <!-- TODO: Link --> the common naming convention. If it doesn't, you may want to consider renaming that ID at this point. Don't worry, this won't hurt consumers in the old frontend system since the ID is mostly used for debugging purposes there. In the new system, it's much more important and appears in app-config files and similar.
+
+Note at the top of the file that it uses the updated import from `@backstage/frontend-plugin-api` that we migrated in the previous section, instead of the old `@backstage/core-plugin-api`.
+
+## Plugin package changes
+
+Now let's turn to the main plugin package where the plugin itself is exported. You will probably already have a `createPlugin` call in here. Before we changed the `core-plugin-api` imports it'll have looked somewhat similar to the following:
+
+```tsx title="in @internal/plugin-example, NOTE THIS IS LEGACY CODE"
+import {
+  storageApiRef,
+  createPlugin,
+  createApiFactory,
+} from '@backstage/core-plugin-api';
+import { workApiRef } from '@internal/plugin-example-react';
+import { WorkImpl } from './WorkImpl';
+
+const workApi = createApiFactory({
+  api: workApiRef,
+  deps: { storageApi: storageApiRef },
+  factory: ({ storageApi }) => new WorkImpl({ storageApi }),
+});
+
+/** @public */
+export const catalogPlugin = createPlugin({
+  id: 'example',
+  apis: [workApi],
+});
+```
+
+The major changes we'll make are
+
+- Change the import to the new package as per the top section of this guide
+- Wrap the existing API factory in a `createApiExtension`
+- Change to the new version of `createPlugin` which exports this extension
+- Change the plugin export to be the default instead
+
+```tsx title="in @internal/plugin-example"
+import {
+  storageApiRef,
+  createPlugin,
+  createApiFactory,
+  createApiExtension,
+} from '@backstage/frontend-plugin-api';
+import { workApiRef } from '@internal/plugin-example-react';
+import { WorkImpl } from './WorkImpl';
+
+const workApi = createApiExtension({
+  api: workApiRef,
+  factory: () =>
```

Comment:
Again the callback form here is unnecessarily verbose and we can stick to providing the factory directly.